shards: reduce blocking in locking strategy

tsenart opened the pull request:
This commit is an attempt to reduce the blocking that occurs when `multiScheduler.Acquire` [calls pile up](https://github.com/sourcegraph/sourcegraph/issues/25872#issuecomment-943311299) after a `multiScheduler.Exclusive` call happens, waiting for already running searches to finish AND the work protected by the Exclusive process.

To that end, we:

1. Remove the `Exclusive` lock functionality from `multiScheduler` which now concerns itself only with controlling concurrent searches and their fairness (interactive + batch).
2. Make `shardedSearcher.replace` work with batches to amortize the initial loading sorting cost instead of sorting it on the search path.
3. Leverage `runtime.SetFinalizer` to call `Close` on a `*rankedSearcher` that was dropped when it is no longer referenced anywhere by any on-going searches.

Part of https://github.com/sourcegraph/sourcegraph/issues/25872

Commit:
shards: reduce blocking in locking strategy

This commit is an attempt to reduce the blocking that occurs when
multiScheduler.Acquire calls pile up after a multiScheduler.Exclusive call happens, waiting
for already running searches to finish AND the work protected by the Exclusive process.

To that end, we:

1. Remove the `Exclusive` lock functionality from `multiScheduler` which
   now concerns itself only with controlling concurrent searches and
   their fairness (interactive + batch).
2. Incrementally update a `rankedShards` heap in `rankedSearcher` on
   every `replace` call, rather than lazily doing it on the search path.
3. Wrap a `rankedSearcher` with a `sync.RWMutex` that makes its `Close`
   operation block until all `Search` and `List` calls on it finish. We
   call `Close` after removing that `*rankedShard` from the heap and
   map.

Part of https://github.com/sourcegraph/sourcegraph/issues/25872

tsenart commented:
I'd like to pair review this tomorrow and do some live end-to-end testing @keegancsmith. Meanwhile @rmmh PTAL!

Commit:
fixup! Add test for ranking

tsenart commented:
@rmmh: Turning this into draft for now. While the ranked ordering is easy to fix, @keegancsmith pointed out that we have a race in the `rankedShard` Close operation — The current code doesn't guaranteed new searches won't start after calling Close() in replace, so we're thinking about solutions to that and will pair on Monday. A slow and bad fix would be to acquire that rankedShard lock on each shard returned from getShards.

Commit:
wip move from heap to sorted slice

Commit:
shards: Add  BenchmarkRLock

Commit:
ignore os.ErrClosed errors

Commit:
fixup! batch loading

tsenart commented:
For Monday, got a weird fault error running tests with `-race`:

```
oekt:ts/reduce-blocking-in-locking-strategy* λ go test -count=1 -race ./...
ok  	github.com/google/zoekt	78.834s
unexpected fault address 0xa1fa029
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa1fa029 pc=0x10ad329]

goroutine 176 [running]:
runtime.throw({0x170c78c, 0x10adad9})
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/panic.go:1198 +0x71 fp=0xc0001338b0 sp=0xc000133880 pc=0x1077771
runtime.sigpanic()
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/signal_unix.go:742 +0x2f6 fp=0xc000133900 sp=0xc0001338b0 pc=0x108f436
runtime.memmove()
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/memmove_amd64.s:181 +0x129 fp=0xc000133908 sp=0xc000133900 pc=0x10ad329
runtime.slicecopy(0xc00032d328, 0x10ed1e5, 0xa1fa029, 0x10ed0b6, 0x10000c00034e670)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/slice.go:314 +0xd1 fp=0xc000133950 sp=0xc000133908 pc=0x1090c51
github.com/google/zoekt/shards.copySlice(...)
	/Users/tomas/Code/zoekt/shards/shards.go:710
github.com/google/zoekt/shards.copyFiles(0xc0000cc240)
	/Users/tomas/Code/zoekt/shards/shards.go:718 +0x285 fp=0xc000133a08 sp=0xc000133950 pc=0x160e6c5
github.com/google/zoekt/shards.(*shardedSearcher).Search(0xc00015c0a0, {0x17d4ac0, 0xc0001f7ce0}, {0x17c94e0, 0xc00000ea80}, 0xc00015c320)
	/Users/tomas/Code/zoekt/shards/shards.go:551 +0x7d9 fp=0xc000133c40 sp=0xc000133a08 pc=0x160ad59
github.com/google/zoekt/shards.(*directorySearcher).Search(0xc00000ea68, {0x17d4ac0, 0xc0001f7ce0}, {0x17c94e0, 0xc00000ea80}, 0x1)
	<autogenerated>:1 +0x9c fp=0xc000133ca8 sp=0xc000133c40 pc=0x1617d3c
github.com/google/zoekt/shards.(*typeRepoSearcher).Search(0xc00034e360, {0x17d4a50, 0xc0000280f8}, {0x17c94e0, 0xc00000ea80}, 0xc00015c320)
	/Users/tomas/Code/zoekt/shards/eval.go:39 +0x30d fp=0xc000133db0 sp=0xc000133ca8 pc=0x160420d
github.com/google/zoekt/build.TestBasic.func1.1(0x1734cb0)
	/Users/tomas/Code/zoekt/build/e2e_test.go:135 +0xfe fp=0xc000133f48 sp=0xc000133db0 pc=0x1627e1e
github.com/google/zoekt/build.retryTest.func1()
	/Users/tomas/Code/zoekt/build/e2e_test.go:168 +0x9a fp=0xc000133fe0 sp=0xc000133f48 pc=0x1628bda
runtime.goexit()
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc000133fe8 sp=0xc000133fe0 pc=0x10ac3a1
created by github.com/google/zoekt/build.retryTest
	/Users/tomas/Code/zoekt/build/e2e_test.go:165 +0x2c5

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00029b6c0, {0x171000c, 0x9}, 0x1734c40)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1307 +0x752
testing.runTests.func1(0x0)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1598 +0x9a
testing.tRunner(0xc00029b6c0, 0xc0001c9b40)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1259 +0x230
testing.runTests(0xc000284380, {0x1b3a2e0, 0xe, 0xe}, {0xa, 0x1, 0x1b45460})
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1596 +0x7cb
testing.(*M).Run(0xc000284380)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1504 +0x9d2
github.com/google/zoekt/build.TestMain(0x480)
	/Users/tomas/Code/zoekt/build/builder_test.go:203 +0x14f
main.main()
	_testmain.go:71 +0x265

goroutine 46 [runnable]:
syscall.syscall(0x10c8fa0, 0xba04080, 0xc0000e6000, 0xc0000e8000)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/sys_darwin.go:22 +0x3b
syscall.readdir_r(0x166a840, 0xc0000e6000, 0xc0000e8000)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/syscall/zsyscall_darwin_amd64.go:1186 +0x57
os.(*File).readdir(0xc000f4e028, 0xffffffffffffffff, 0x0)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/os/dir_darwin.go:48 +0x4dc
os.(*File).Readdirnames(0xc000f4e028, 0x48)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/os/dir.go:70 +0x45
path/filepath.glob({0xc00030c5a0, 0x48}, {0xc00030c5e9, 0x7}, {0x0, 0x0, 0x0})
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/path/filepath/match.go:337 +0x16d
path/filepath.Glob({0xc00030c5a0, 0x2})
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/path/filepath/match.go:264 +0x365
github.com/google/zoekt/shards.(*DirectoryWatcher).scan(0xc00015c190)
	/Users/tomas/Code/zoekt/shards/watcher.go:103 +0xb1
github.com/google/zoekt/shards.(*DirectoryWatcher).watch.func2()
	/Users/tomas/Code/zoekt/shards/watcher.go:231 +0xbf
created by github.com/google/zoekt/shards.(*DirectoryWatcher).watch
	/Users/tomas/Code/zoekt/shards/watcher.go:228 +0x23d

goroutine 45 [select]:
github.com/google/zoekt/shards.(*DirectoryWatcher).watch.func1()
	/Users/tomas/Code/zoekt/shards/watcher.go:208 +0x15c
created by github.com/google/zoekt/shards.(*DirectoryWatcher).watch
	/Users/tomas/Code/zoekt/shards/watcher.go:206 +0x193

goroutine 44 [syscall]:
syscall.syscall6(0x1600ca0, 0x6, 0x0, 0x0, 0xc00006be70, 0xa, 0x1b75640)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/runtime/sys_darwin.go:44 +0x3b
golang.org/x/sys/unix.kevent(0xc00031f790, 0x0, 0x5, 0xc00006be70, 0x4, 0x1b75640)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/golang.org/x/sys@v0.0.0-20210630005230-0f9fa26af87c/unix/zsyscall_darwin_amd64.go:276 +0xb7
golang.org/x/sys/unix.Kevent(0x10aef97, {0x0, 0x0, 0x0}, {0xc00006be70, 0xa, 0x48}, 0xc0002f0749)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/golang.org/x/sys@v0.0.0-20210630005230-0f9fa26af87c/unix/syscall_bsd.go:429 +0x87
github.com/fsnotify/fsnotify.read(0xc0000bcc60, {0xc00006be70, 0xa, 0xa}, 0x5d)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/github.com/fsnotify/fsnotify@v1.4.9/kqueue.go:511 +0x73
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0000bcc60)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/github.com/fsnotify/fsnotify@v1.4.9/kqueue.go:274 +0xd1
created by github.com/fsnotify/fsnotify.NewWatcher
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/github.com/fsnotify/fsnotify@v1.4.9/kqueue.go:62 +0x3c5

goroutine 99 [chan receive]:
gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun(0xc000035500)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/gopkg.in/natefinch/lumberjack.v2@v2.0.0/lumberjack.go:379 +0x65
created by gopkg.in/natefinch/lumberjack%2ev2.(*Logger).mill.func1
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/gopkg.in/natefinch/lumberjack.v2@v2.0.0/lumberjack.go:390 +0xf1

goroutine 75 [chan receive]:
github.com/google/zoekt/ctags.(*lockedParser).Parse.func1()
	/Users/tomas/Code/zoekt/ctags/json.go:75 +0x1fb
created by github.com/google/zoekt/ctags.(*lockedParser).Parse
	/Users/tomas/Code/zoekt/ctags/json.go:73 +0x2de

goroutine 115 [chan receive]:
testing.(*T).Run(0xc0003129c0, {0x17104ef, 0x9}, 0xc0000bcfc0)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1307 +0x752
github.com/google/zoekt/build.TestBasic(0xc0003129c0)
	/Users/tomas/Code/zoekt/build/e2e_test.go:112 +0xe1a
testing.tRunner(0xc0003129c0, 0x1734c40)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1306 +0x727

goroutine 41 [chan receive]:
gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun(0xc0000344e0)
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/gopkg.in/natefinch/lumberjack.v2@v2.0.0/lumberjack.go:379 +0x65
created by gopkg.in/natefinch/lumberjack%2ev2.(*Logger).mill.func1
	/Users/tomas/.asdf/installs/golang/1.17.1/packages/pkg/mod/gopkg.in/natefinch/lumberjack.v2@v2.0.0/lumberjack.go:390 +0xf1

goroutine 175 [chan receive]:
github.com/google/zoekt/build.retryTest(0xc000103d40, 0xc0000bd020)
	/Users/tomas/Code/zoekt/build/e2e_test.go:175 +0x2d4
github.com/google/zoekt/build.TestBasic.func1(0xc000103d40)
	/Users/tomas/Code/zoekt/build/e2e_test.go:115 +0x23f
testing.tRunner(0xc000103d40, 0xc0000bcfc0)
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/Users/tomas/.asdf/installs/golang/1.17.1/go/src/testing/testing.go:1306 +0x727
FAIL	github.com/google/zoekt/build	6.149s
?   	github.com/google/zoekt/cmd	[no test files]
?   	github.com/google/zoekt/cmd/zoekt	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-archive-index	14.903s
?   	github.com/google/zoekt/cmd/zoekt-git-clone	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-git-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-hg-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-indexserver	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-merge-index	3.587s
?   	github.com/google/zoekt/cmd/zoekt-mirror-bitbucket-server	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gerrit	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-github	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gitiles	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gitlab	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-repo-index	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-sourcegraph-indexserver	58.760s
?   	github.com/google/zoekt/cmd/zoekt-test	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-webserver	[no test files]
ok  	github.com/google/zoekt/ctags	0.410s
?   	github.com/google/zoekt/debugserver	[no test files]
ok  	github.com/google/zoekt/gitindex	28.587s
ok  	github.com/google/zoekt/ignore	0.281s
?   	github.com/google/zoekt/internal/mockSearcher	[no test files]
ok  	github.com/google/zoekt/query	0.599s
ok  	github.com/google/zoekt/rpc	0.300s
?   	github.com/google/zoekt/rpc/internal/srv	[no test files]
ok  	github.com/google/zoekt/shards	29.337s
ok  	github.com/google/zoekt/stream	0.314s
?   	github.com/google/zoekt/trace	[no test files]
ok  	github.com/google/zoekt/web	17.241s
FAIL
zoekt:ts/reduce-blocking-in-locking-strategy* λ 
```

rmmh commented:
The overall idea is to remove the global searcher RWLock that's held during searches in favor of more granular locks for each shard, right?

One more idea is to use atomic pointer replacement to store a slice of ordered shards that's updated in replace/replaceBatch, removing the need for the read/search path to take a lock on the searcher at all.

Commit:
fixup! use runtime.SetFinalizer

Commit:
fixup! Fix flake

Commit:
Update shards/shards.go

Commit:
fixup! remove BenchmarkRLock

tsenart commented:
> The overall idea is to remove the global searcher RWLock that's held during searches in favor of more granular locks for each shard, right?

Yes, because it was blocking searches for more than 1s in the scenario explained in the PR description. 
 
> One more idea is to use atomic pointer replacement to store a slice of ordered shards that's updated in replace/replaceBatch, removing the need for the read/search path to take a lock on the searcher at all.

That's what the current state of this PR does, while relying on `runtime.SetFinalizer` to call `Close` on the underlying mmaped file of dropped shards once there are no more on-going searches referencing it.

PTAL! Gonna get back to this with @keegancsmith on Monday morning. Signing-off for the weekend now.

Commit:
fixup! address review feedback

keegancsmith committed:
attempt at using io.Closer in streamSearch

keegancsmith committed:
move copyFiles back into place

We use KeepAlive to signal we are still using the shards. Using
copyFiles inside of streamSearch is a regression for Search since it
will greatly increase the number of copies. We still have uses of Search
in Sourcegraph.

keegancsmith committed:
fixup! move copyFiles back into place

keegancsmith committed:
fix help on batch replace metric

keegancsmith committed:
remove close duration metric

No longer important to track since it should be quick and doesn't block
anything.

keegancsmith committed:
add a comment about SetFinalizer

keegancsmith committed:
handle s.ranked not being set

keegancsmith committed:
shardedSearcher.Close reuses replace

Safer way for us to call Close than before.
